From 95843d7978f2e4311abfdf756b68d28a48e1d7fe Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:42 -0400
Subject: [fs] jbd2: minimize region locked by j_list_lock in journal_get_create_access()

Message-id: <1414058125-4183-129-git-send-email-lczerner@redhat.com>
Patchwork-id: 98200
O-Subject: [RHEL 7.1 PATCH v2 128/171] jbd2: minimize region locked by j_list_lock in journal_get_create_access()
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Theodore Ts'o <tytso@mit.edu>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 6e4862a5bb9d12be87e4ea5d9a60836ebed71d28

 It's not needed until we start trying to modifying fields in the
 journal_head which are protected by j_list_lock.

 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/jbd2/transaction.c b/fs/jbd2/transaction.c
index 902b99b..71bd0e4 100644
--- a/fs/jbd2/transaction.c
+++ b/fs/jbd2/transaction.c
@@ -1071,7 +1071,6 @@ int jbd2_journal_get_create_access(handle_t *handle, struct buffer_head *bh)
   * reused here.
   */
  jbd_lock_bh_state(bh);
- spin_lock(&journal->j_list_lock);
  J_ASSERT_JH(jh, (jh->b_transaction == transaction ||
   jh->b_transaction == NULL ||
   (jh->b_transaction == journal->j_committing_transaction &&
@@ -1094,12 +1093,14 @@ int jbd2_journal_get_create_access(handle_t *handle, struct buffer_head *bh)
   jh->b_modified = 0;
 
   JBUFFER_TRACE(jh, "file as BJ_Reserved");
+  spin_lock(&journal->j_list_lock);
   __jbd2_journal_file_buffer(jh, transaction, BJ_Reserved);
  } else if (jh->b_transaction == journal->j_committing_transaction) {
   /* first access by this transaction */
   jh->b_modified = 0;
 
   JBUFFER_TRACE(jh, "set next transaction");
+  spin_lock(&journal->j_list_lock);
   jh->b_next_transaction = transaction;
  }
  spin_unlock(&journal->j_list_lock);
-- 
1.7.1